File: app/src/main/java/com/flatstack/android/utils/ui/BaseDialogFragment.java
package com.flatstack.android.utils.ui;

import android.app.Dialog;
import android.content.Context;
import android.os.Bundle;
import android.support.annotation.LayoutRes;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.DialogFragment;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentActivity;
import android.support.v4.app.FragmentTransaction;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.Window;

import butterknife.ButterKnife;

public abstract class BaseDialogFragment extends DialogFragment {

    @LayoutRes public abstract int getLayoutRes();

    protected static <T extends BaseDialogFragment> T show(@NonNull T dialogFragment,
                                                           @NonNull FragmentActivity activity) {
        FragmentTransaction ft = activity.getSupportFragmentManager().beginTransaction();
        Fragment prev = activity.getSupportFragmentManager()
            .findFragmentByTag(dialogFragment.getClass().getName());
        if (prev != null) {
            ft.remove(prev);
            DialogFragment df = (DialogFragment) prev;
            df.dismissAllowingStateLoss();
        }
        ft.addToBackStack(null);
        dialogFragment.show(ft, dialogFragment.getClass().getName());
        return dialogFragment;
    }

    @Override public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
            parseArguments(getArguments());
        }
        if (savedInstanceState != null) {
            restoreState(savedInstanceState);
        }
    }

    protected void restoreState(@NonNull Bundle savedState) {
    }

    protected void parseArguments(@NonNull Bundle args) {
        throw new IllegalStateException("should be overridden");
    }

    @NonNull @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        Dialog dialog = super.onCreateDialog(savedInstanceState);
        dialog.getWindow().requestFeature(Window.FEATURE_NO_TITLE);
        return dialog;
    }

    @Override
    public View onCreateView(LayoutInflater inflater,
                             @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        View v = inflater.inflate(getLayoutRes(), container, false);
        ButterKnife.bind(this, v);
        return v;
    }

    @Override public void onDestroyView() {
        ButterKnife.unbind(this);
        super.onDestroyView();
    }

    public Context context() {
        return getActivity();
    }

    public FragmentActivity activity() {
        return getActivity();
    }
}

File: app/src/main/java/com/flatstack/android/utils/ui/BaseDialogFragment.kt
package com.flatstack.android.utils.ui

import android.app.Dialog
import android.os.Bundle
import android.support.v4.app.DialogFragment
import android.support.v4.app.FragmentActivity
import android.view.LayoutInflater
import android.view.ViewGroup
import android.view.Window

abstract class BaseDialogFragment : DialogFragment() {
    internal abstract val layoutRes: Int

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        arguments?.let { parseArguments(it) }
        savedInstanceState?.let { restoreState(it) }
    }

    protected fun restoreState(savedState: Bundle) {}

    protected open fun parseArguments(args: Bundle) {
        throw IllegalStateException("should be overridden")
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val dialog = super.onCreateDialog(savedInstanceState)
        dialog.window?.requestFeature(Window.FEATURE_NO_TITLE)
        return dialog
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ) = inflater.inflate(layoutRes, container, false)

    companion object {
        protected fun <T : BaseDialogFragment> show(
            dialogFragment: T,
            activity: FragmentActivity
        ): T {
            val ft = activity.supportFragmentManager.beginTransaction()
            val prev = activity.supportFragmentManager
                    .findFragmentByTag(dialogFragment.javaClass.name)
            prev?.let {
                ft.remove(it)
                val df = prev as DialogFragment
                df.dismissAllowingStateLoss()
            }
            ft.addToBackStack(null)
            dialogFragment.show(ft, dialogFragment.javaClass.name)
            return dialogFragment
        }
    }
}
